src/components/Table/hooks.tsx

```diff
@@ -0,0 +1,258 @@
+        return (
+          row.cells.filter((cell) => {
+            if (cell.value.includes(action.searchString)) {
+              return true;
+            }
+            return false;
+          }).length > 0
+        );
+      });
+      return stateCopySearch;
+    }
+    default:
+      throw new Error("Invalid reducer action");
+  }
+};
+
+export const useTable = <T extends DataType>(
+  columns: ColumnType<T>[],
+  data: T[],
+  options?: UseTableOptionsType<T>
+): UseTableReturnType<T> => {
+  const columnsWithSorting: ColumnStateType<T>[] = useMemo(
+    () =>
+      columns.map((column) => {
+        return {
+          ...column,
+          label: column.label ? column.label : column.name,
+          hidden: column.hidden ? column.hidden : false,
+          sort: column.sort,
+          sorted: {
+            on: false,
+          },
+        };
+      }),
+    [columns]
+  );
+  const columnsByName = useMemo(() => getColumnsByName(columnsWithSorting), [columnsWithSorting]);
+
+  const tableData: RowType<T>[] = useMemo(() => {
+    const sortedData = sortDataInOrder(data, columnsWithSorting);
+
+    const newData = sortedData.map((row, idx) => {
+      return {
+        id: idx,
+        selected: false,
+        hidden: false,
+        original: row,
+        cells: Object.entries(row)
+          .map(([column, value]) => {
+            return {
+              hidden: columnsByName[column].hidden,
+              field: column,
+              value,
+              render: makeRender(value, columnsByName[column].render, row),
+            };
+          })
+          .filter((cell) => !cell.hidden),
+      };
+    });
+    return newData;
+  }, [data, columnsWithSorting, columnsByName]);
+
+  const reducer = createReducer<T>();
+
+  const [state, dispatch] = useReducer(reducer, {
+    columns: columnsWithSorting,
+    columnsByName,
+    originalRows: tableData,
+    rows: tableData,
+    selectedRows: [],
+    toggleAllState: false,
+    filterOn: !!options?.filter,
+    sortColumn: null,
+    paginationEnabled: !!options?.pagination,
+    pagination: {
+      page: 1,
+      perPage: 10,
+      canNext: true,
+      canPrev: false,
+      nextPage: () => {
+        // nextPage feature
+      },
+      prevPage: () => {
+        // prevPage feature
+      },
+    },
+  });
+
+  useEffect(() => {
+    dispatch({ type: "SET_ROWS", data: tableData });
+  }, [tableData]);
+
+  const headers: HeaderType<T>[] = useMemo(() => {
+    return [
+      ...state.columns.map((column) => {
+        const label = column.label ? column.label : column.name;
+        return {
+          ...column,
+          render: makeHeaderRender(label, column.headerRender),
+        };
+      }),
+    ];
+  }, [state.columns]);
+
+  useEffect(() => {
+    if (options && options.filter) {
+      dispatch({ type: "GLOBAL_FILTER", filter: options.filter });
+    }
+  }, [options?.filter, options]);
```

Comment:
You don't need `options?.filter` as `options` is already a dependency. Please make sure this doesn't re-render too much.

Reply:
Ok